units: 's / m' and 's / m' are not convertible

To paraphrase from https://github.com/PlasmaPy/PlasmaPy/issues/587, exceptions like `astropy.units.core.UnitConversionError: 's / m' and 's / m' are not convertible` (and a few other cases with first, second and third powers of second over meter, curiously) started popping up in our travis tests as seen at https://travis-ci.org/PlasmaPy/PlasmaPy/jobs/466396211 .



For a brief overview, running `python setup.py test -d` and getting into `plasmapy.physics.distribution.Maxwellian_1D`:

```python
(Pdb) distFunc
<Quantity 5.91632969e-07 s / m>
(Pdb) p u.s / u.m
Unit("s / m")
(Pdb) p distFunc
<Quantity 5.91632969e-07 s / m>
(Pdb) p distFunc.unit
Unit("s / m")
(Pdb) p distFunc.to(u.s / u.m)
*** astropy.units.core.UnitConversionError: 's / m' and 's / m' are not convertible
```

 I've managed to figure out that this is localized to `astropy` 3.1. As suggested by @astrofrog on slack, I ran `git bisect` with `pytest --doctest-modules $PLASMAPY/plasmapy/physics/distribution.py` and I was able to identify the first bad commit as
```git
3a478ca29e86144d6c5a0305dde86169a647ff63 is the first bad commit
commit 3a478ca29e86144d6c5a0305dde86169a647ff63
Author: Marten van Kerkwijk <mhvk@astro.utoronto.ca>
Date:   Thu Jul 12 21:01:28 2018 -0700

    Further short-cuts for single-unit CompositeUnit initialization.

:040000 040000 1ba7cb9440215ab641197f91ef1623ad9bb39378 5c02db4ede9146ec12f84f57a0382f4900777544 M      astropy
```

I'm currently trying to find out the exact cause of this bug in that commit, but I figured it's best to submit this issue here before I accidentally close this tab or something.

Here's a more detailed traceback done from 3a478ca2:

```python
plasmapy/physics/tests/test_distribution.py:21 (test_astropy)
def test_astropy():
        v=1*u.m/u.s
>       Maxwellian_1D(v=v, T=30000 * u.K, particle='e', v_drift=0 * u.m / u.s)

plasmapy/physics/tests/test_distribution.py:24: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
plasmapy/physics/distribution.py:142: in Maxwellian_1D
    return distFunc.to(u.s / u.m)
../../astropy/astropy/units/quantity.py:669: in to
    return self._new_view(self._to_value(unit, equivalencies), unit)
../../astropy/astropy/units/quantity.py:641: in _to_value
    equivalencies=equivalencies)
../../astropy/astropy/units/core.py:984: in to
    return self._get_converter(other, equivalencies=equivalencies)(value)
../../astropy/astropy/units/core.py:915: in _get_converter
    raise exc
../../astropy/astropy/units/core.py:901: in _get_converter
    self, other, self._normalize_equivalencies(equivalencies))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Unit("s / m"), unit = Unit("s / m"), other = Unit("s / m")
equivalencies = []

    def _apply_equivalencies(self, unit, other, equivalencies):
        """
        Internal function (used from `_get_converter`) to apply
        equivalence pairs.
        """
        def make_converter(scale1, func, scale2):
            def convert(v):
                return func(_condition_arg(v) / scale1) * scale2
            return convert
    
        for funit, tunit, a, b in equivalencies:
            if tunit is None:
                try:
                    ratio_in_funit = (other.decompose() /
                                      unit.decompose()).decompose([funit])
                    return make_converter(ratio_in_funit.scale, a, 1.)
                except UnitsError:
                    pass
            else:
                try:
                    scale1 = funit._to(unit)
                    scale2 = tunit._to(other)
                    return make_converter(scale1, a, scale2)
                except UnitsError:
                    pass
                try:
                    scale1 = tunit._to(unit)
                    scale2 = funit._to(other)
                    return make_converter(scale1, b, scale2)
                except UnitsError:
                    pass
    
        def get_err_str(unit):
            unit_str = unit.to_string('unscaled')
            physical_type = unit.physical_type
            if physical_type != 'unknown':
                unit_str = "'{0}' ({1})".format(
                    unit_str, physical_type)
            else:
                unit_str = "'{0}'".format(unit_str)
            return unit_str
    
        unit_str = get_err_str(unit)
        other_str = get_err_str(other)
    
        raise UnitConversionError(
            "{0} and {1} are not convertible".format(
>               unit_str, other_str))
E       astropy.units.core.UnitConversionError: 's / m' and 's / m' are not convertible

../../astropy/astropy/units/core.py:885: UnitConversionError
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -300,6 +300,8 @@ astropy.uncertainty
 astropy.units
 ^^^^^^^^^^^^^
 
+- Ensure correctness of units when raising to a negative power. [#8263]
+
 astropy.utils
 ^^^^^^^^^^^^^
 

--- a/astropy/units/core.py
+++ b/astropy/units/core.py
@@ -2022,8 +2022,10 @@ class CompositeUnit(UnitBase):
                         "bases must be sequence of UnitBase instances")
             powers = [validate_power(p) for p in powers]
 
-        if not decompose and len(bases) == 1:
-            # Short-cut; with one unit there's nothing to expand and gather.
+        if not decompose and len(bases) == 1 and powers[0] >= 0:
+            # Short-cut; with one unit there's nothing to expand and gather,
+            # as that has happened already when creating the unit.  But do only
+            # positive powers, since for negative powers we need to re-sort.
             unit = bases[0]
             power = powers[0]
             if power == 1:
@@ -2038,6 +2040,7 @@ class CompositeUnit(UnitBase):
                 self._bases = unit.bases
                 self._powers = [operator.mul(*resolve_fractions(p, power))
                                 for p in unit.powers]
+
             self._scale = sanitize_scale(scale)
         else:
             # Regular case: use inputs as preliminary scale, bases, and powers,

--- a/astropy/units/tests/test_units.py
+++ b/astropy/units/tests/test_units.py
@@ -807,3 +807,15 @@ def test_unit_summary_prefixes():
             assert prefixes == 'No'
         elif unit.name == 'vox':
             assert prefixes == 'Yes'
+
+
+def test_raise_to_negative_power():
+    """Test that order of bases is changed when raising to negative power.
+
+    Regression test for https://github.com/astropy/astropy/issues/8260
+    """
+    m2s2 = u.m ** 2 / u.s **2
+    spm = m2s2 ** (-1 / 2)
+    assert spm.bases == [u.s, u.m]
+    assert spm.powers == [1, -1]
+    assert spm == u.s / u.m

